NCGenerics: restrict conditional Copyable reqs

kavon opened the pull request:
It doesn't really make sense for a conditional conformance requirement for `Copyable` to depend on any other requirement other than other `Copyable` conformance requirements.

resolves rdar://124967739

kavon committed:
NCGenerics: restrict conditional Copyable reqs

It doesn't really make sense for a conditional conformance requirement
for `Copyable` to depend on any other requirement other than other
`Copyable` conformance requirements.

resolves rdar://124967739

kavon commented:
@swift-ci smoke test

slavapestov commented on lib/Sema/TypeCheckInvertible.cpp:
```diff
@@ -140,10 +140,45 @@ static void checkInvertibleConformanceCommon(DeclContext *dc,
+      //
+      // In theory, it could depend on any invertible protocol, but it may be
+      // confusing if we permitted that and this simplifies the model a bit.
+      for (auto req : condReqs) {
```
Can you also ban requirements where the first type is a DependentMemberType?

slavapestov commented on lib/Sema/TypeCheckInvertible.cpp:
```diff
@@ -140,10 +140,45 @@ static void checkInvertibleConformanceCommon(DeclContext *dc,
+          break;
+        }
+
+        if (illegalSecondTypeStr)
```
Nitpick: Add { } around multi-line if body

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test linux

kavon committed:
Diagnostics: `noformat` modifier for Type

Sometimes, quotes or an "aka" for a type are not desirable in a
diagnostic, such as one describing a Requirement. This modifier
suppresses such additional formatting a Type in a diagnostic.

kavon commented:
@swift-ci smoke test
